### PR TITLE
WIP: chore(deps): upgrade pgx to v5

### PR DESCRIPTION
**Pull request by janisz**
## Description

> pgx follows semantic versioning for the documented public API on stable releases. 
> [**v5 is the latest stable major version**.](https://github.com/jackc/pgx#version-policy)

- https://github.com/go-gorm/postgres/issues/139
- https://github.com/stackrox/stackrox/pull/4676

Important changes in v5:
- `CommandTag` is a type
- [handling arrays – we have code that convert `[]TextArray` but now it's just a `[]string` so only filtering part is needed.](https://github.com/jackc/pgx/blob/v5.4.2/CHANGELOG.md#array-types)
- handling null values in jsonb there was `'null'` but now it's `null` so I added a change to use `{}` with `pgutils.EmptyOrMap`

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

CI

### CODE REVIEW

**Review by stehessel — CHANGES_REQUESTED** [review]
I think you need some more fixes in the unit tests because of this change:
> CommandTag is now an opaque type instead of directly exposing an underlying []byte.

See for example
```
=== RUN   TestAlertDataStore/TestConnExec
    conn_test.go:106: 
        	Error Trace:	/__w/stackrox/stackrox/pkg/postgres/tests/conn_test.go:106
        	Error:      	Expected nil, but got: pgconn.CommandTag{s:""}
        	Test:       	TestAlertDataStore/TestConnExec
    conn_test.go:112: 
        	Error Trace:	/__w/stackrox/stackrox/pkg/postgres/tests/conn_test.go:112
        	Error:      	Expected nil, but got: pgconn.CommandTag{s:""}
        	Test:       	TestAlertDataStore/TestConnExec
```

**Comment by janisz** [comment]
Indeed the handling or arrays also changed causing problems https://github.com/jackc/pgx/blob/v5.4.2/CHANGELOG.md?plain=1

**Comment by janisz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6766** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6766" target="_blank">Graphite</a> 
    * **PR #6767** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6767" target="_blank">Graphite</a> 
      * **PR #6768** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6768" target="_blank">Graphite</a> 
        * **PR #6769** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6769" target="_blank">Graphite</a> 
          * **PR #6770** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6770" target="_blank">Graphite</a> 
            * **PR #6771** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6771" target="_blank">Graphite</a> 
              * **PR #6772** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6772" target="_blank">Graphite</a> 
                * **PR #6773** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6773" target="_blank">Graphite</a> 
                  * **PR #6774** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6774" target="_blank">Graphite</a> 
                    * **PR #6775** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6775" target="_blank">Graphite</a> 
                      * **PR #6776** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6776" target="_blank">Graphite</a> 
                        * **PR #6777** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6777" target="_blank">Graphite</a> 
                          * **PR #6778** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6778" target="_blank">Graphite</a> 
                            * **PR #6881** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6881" target="_blank">Graphite</a> 
                              * **PR #6882** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6882" target="_blank">Graphite</a> 
                                * **PR #6915** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6915" target="_blank">Graphite</a>  👈
                                  * **PR #6916** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6916" target="_blank">Graphite</a> 
                                * **PR #6894** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6894" target="_blank">Graphite</a> 
                                  * **PR #6907** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6907" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/6915?utm_source=stack-comment).

**Comment by janisz** [comment]
Tests are passing locally but not on CI 🤔 

```
CI=true go test -tags sql_integration  ./pkg/search/postgres/...              
?   	github.com/stackrox/rox/pkg/search/postgres/aggregatefunc	[no test files]
?   	github.com/stackrox/rox/pkg/search/postgres/mapping	[no test files]
ok  	github.com/stackrox/rox/pkg/search/postgres	22.543s
ok  	github.com/stackrox/rox/pkg/search/postgres/query	0.057s
ok  	github.com/stackrox/rox/pkg/search/postgres/query/test	0.201s
ok  	github.com/stackrox/rox/pkg/search/postgres/query/test/uuid	0.239s
```

File: go.mod [hunk 1]
```diff
@@ -249,11 +247,13 @@ require (
...
 	github.com/itchyny/timefmt-go v0.1.5 // indirect
 	github.com/jackc/chunkreader/v2 v2.0.1 // indirect
+	github.com/jackc/pgconn v1.14.0 // indirect
 	github.com/jackc/pgio v1.0.0 // indirect
 	github.com/jackc/pgpassfile v1.0.0 // indirect
+	github.com/jackc/pgproto3/v2 v2.3.2 // indirect
 	github.com/jackc/pgservicefile v0.0.0-20221227161230-091c0ba34f0a // indirect
-	github.com/jackc/pgx/v5 v5.3.1 // indirect
-	github.com/jackc/puddle v1.3.0 // indirect
+	github.com/jackc/pgx/v4 v4.18.0 // indirect
```
Comment (janisz) [inline go.mod:255]: This will be removed in 
- #6916

File: pkg/search/postgres/query/array_query.go [hunk 2]
```diff
@@ -22,15 +21,15 @@ func getStringArrayPostTransformFunc(entry *QueryEntry) (func(val interface{}) i
 		return nil, errors.New("no filter func found")
 	}
 	return func(val interface{}) interface{} {
-		textArray, _ := val.(*pgtype.TextArray)
+		textArray, _ := val.(*[]string)
```
Comment (janisz) [inline pkg/search/postgres/query/array_query.go:24]: 👀

File: tools/generate-helpers/pg-table-bindings/store.go.tpl [hunk 3]
```diff
@@ -205,6 +205,8 @@ func {{ template "insertFunctionName" $schema }}({{ if eq (len $schema.Children)
         pgutils.NilOrTime({{$field.Getter "obj"}}),
         {{- else if eq $field.SQLType "uuid" }}
         pgutils.NilOrUUID({{$field.Getter "obj"}}),
+        {{- else if eq $field.DataType "map" }}
+        pgutils.EmptyOrMap({{$field.Getter "obj"}}),
```
Comment (janisz) [inline tools/generate-helpers/pg-table-bindings/store.go.tpl:209]: Since there was a change in how maps are serialized (`null` vs `'null'`) I've decided to save them as empty object `{}` no matter if they are null or empty.

**Review by janisz — COMMENTED** [review]
Added comments to changes that are not `v4` → `v5` or `nil` → `CommandTag{}`

File: pkg/search/postgres/query/array_query.go [hunk 4]
```diff
@@ -22,15 +21,15 @@ func getStringArrayPostTransformFunc(entry *QueryEntry) (func(val interface{}) i
...
 		if textArray == nil {
-			return (*pgtype.TextArray)(nil)
+			return nil
```
Comment (stehessel) [inline pkg/search/postgres/query/array_query.go:26]: Why did you change the return value from typed nil to untyped nil? This changes the behavior if the return value is nil checked.

File: tools/generate-helpers/pg-table-bindings/store.go.tpl (hunk 3 above)
Comment (janisz) [reply tools/generate-helpers/pg-table-bindings/store.go.tpl]: Extracted 
- https://github.com/stackrox/stackrox/pull/6929

File: pkg/search/postgres/query/array_query.go (hunk 4 above)
Comment (janisz) [reply pkg/search/postgres/query/array_query.go]: Good point, fixed

File: pkg/search/postgres/query/array_query.go (hunk 2 above)
Comment (stehessel) [inline pkg/search/postgres/query/array_query.go:24]: Looking at the pgx changelog, it seems that arrays are now generic types. Did you consider using https://github.com/jackc/pgx/blob/038fc448c1cc91b9372a6a0d151789f41536f033/pgtype/array.go#L379 instead of `[]string`?

Comment (janisz) [reply pkg/search/postgres/query/array_query.go:24]: Good point. Since the `TestSearchComparison/TestDeploymentSearchResults/test#02` was failing and when I attached debugger to it shows me that `val` is `[]string` so I did not try `pgtype.Array` until now and it failed as conversion will not work.

**Comment by dashrews78** [comment]
I haven't had a chance to look at this in detail yet, but I do have a question.  What level of testing have we done on this?  Reason I ask is that this looks like a significant upgrade and we've seen some things in the past where pgx took some liberties with how they do things.  My concern is they adjusted some behaviors and we won't necessarily see them with just vanilla CI.  Things we found during scale and chaos type testing would be missed in CI.

Additionally should we go through the v5 changelog and write tickets to step up to some of their improved flows such as the new way batching works, etc?

**Comment by stehessel** [comment]
> I haven't had a chance to look at this in detail yet, but I do have a question. What level of testing have we done on this? Reason I ask is that this looks like a significant upgrade and we've seen some things in the past where pgx took some liberties with how they do things. My concern is they adjusted some behaviors and we won't necessarily see them with just vanilla CI. Things we found during scale and chaos type testing would be missed in CI.

Agreed, my review was mostly just looking at the changes surface level, but I don't believe we should bump this just because the unit tests pass. I don't know how these upgrades have been done in the past, but I'd be more comfortable with some larger scale tests on the new version.

**Comment by janisz** [comment]
I concur we need some better test plan. So far I only manually verified some generated queries and besides of map nil serialization I haven't seen any differences. But queries are just a small part of what pgx does. I can enable run UI tests as they run more gql queries.

Unfortunately there is no upgrade guide for pgx.  I was under the impression it's mostly backward compatible. E.g.: gorm just merged it and released as bugfix release and we already are using this version.

Maybe better time to merge it will be just after the release so we will have more time to catch possible bugs. We will need to bump it at some point.

**Comment by janisz** [comment]
@connorgorman @mtodor could you help with perf/scale tests?

**Comment by janisz** [comment]
Closing, we can reopen it once we have a proper test plan for this upgrade.
